### ISSUE TITLE
Add SentencePiece binding to torchtext

### ISSUE THREAD

**Issue by zhangguanheng66** [pytorch/text#589]
SentencePiece is an unsupervised text tokenizer that is mainly applicable for text-based systems. The size of vocab is predetermined prior to train the tokenizer. The subword unit implemented in SentencePiece allows for tokenizing raw sentences without language-specific pre/postprocessing.

In addition to text-based system, the SentencePiece may be also applicable for torchaudio script.

@cpuhrsch @vincentqb @soumith

**Comment by zhangguanheng66** [comment]
The first part of work is to apply sentencepiece to some supervised learning datasets (i.e. textclassification datasets). The subword method in sentencepiece should improve the results in the text classification tutorial [here](https://pytorch.org/tutorials/beginner/text_sentiment_ngrams_tutorial.html).

Then, we could apply sentencepiece to unsupervised learning datasets.

**Comment by vincentqb** [comment]
> In addition to text-based system, the SentencePiece may be also applicable for torchaudio script.

[Example](https://github.com/pytorch/audio/tree/master/examples/interactive_asr) from torchaudio

**Comment by zhangguanheng66** [comment]
The TransformerBag plus MultiheadAttentionBag will add more flexibility for the input sequences. For the sequences with different lengths, no padding is required, and mask/offsets are used for the attention mechanism. Though it’s a pleasant to have such kind of capability, it requires some fundamental changes on the current modules (both nn.MultiheadAttention and nn.Transformer ). We have to re-design the modules and re-write them from scratch. We identify it as a nice feature to have and are open to some contributors for the implementation. The existing modules have been already applied for some non-text based sequences (like PySpeech). Users just need to pad the sequences before sending to the current attention mechanism.
